acoular/spectra.py

```diff
@@ -384,17 +399,28 @@ def calc_csm(self):
         numfreq = int(self.block_size / 2 + 1)
         csm_shape = (numfreq, t.numchannels, t.numchannels)
         csm_upper = zeros(csm_shape, dtype=self.precision)
-        # print "num blocks", self.num_blocks
+        # print( "num blocks", self.num_blocks)
         # for backward compatibility
         if self.calib and self.calib.num_mics > 0:
             if self.calib.num_mics == t.numchannels:
                 wind = wind * self.calib.data[newaxis, :]
             else:
                 raise ValueError('Calibration data not compatible: %i, %i' % (self.calib.num_mics, t.numchannels))
         # get time data blockwise
+        i = 0
```

Comment:
better to use `enumerate` in the for statement

Reply:
This is a bit tricky. Using `enumerate` will cause the enumeration to be infinite, which means an overflow at some point. If I then use a `%` or rather a `divmod` operation, the distance between the last multiple of `num_blocks` before and the first after the overflow will not be equal to `num_blocks`, introducing a **latency bug** for long data streams.

Example with a `uint8_t` counter:
```python
num_blocks = 10
for i, data in enumerate(gen):  # assume that i is 8 bit
     q, r = divmod(i, num_blocks)
     process(data)
     if r == 0:
        postprocess()
```
The last time `postprocess()` is called before overflow is at `i = 250`. The next time round it will be called at `i = 10`, which means that between the iterations across the overflow, `process(data)` was called 16 times instead of 10 times. This bug is rare, but it is one nonetheless.

Reply:
I do not agree. `i` as returned by the `enumerate` generator is a Python int. In Python 3 this is unbounded (no upper limit). Even in a hypothetical 'lite' Python implementation this would be something like the processor word size (nowadays at least 16 bits I assume) - this was the way it was in Python 2.

Reply:
The proposed improvements for the `PowerSpectra` class are based on `range` instead ([6af95f6](https://github.com/acoular/acoular/pull/285/commits/6af95f678bd75212994065fd8994607b3258eca2)). However, I will use `enumerate()` in the not yet pushed `PowerSpectraInOut` class.